sound/soc/sof/intel/hda-stream.c

```diff
@@ -88,31 +88,46 @@ static int hda_setup_bdle(struct snd_sof_dev *sdev,
 /*
  * set up Buffer Descriptor List (BDL) for host memory transfer
  * BDL describes the location of the individual buffers and is little endian.
+ *
+ * The typical BDL entries after set up should looks like this:
+ *
+ * bdl[0] from the init_offset,  n entries before wrap,
+ * the e-1 full entries, and the last e entry.
+ *
+ *			     init_offset
+ *				 |
+ * |------|...........|---------S-------|------|........|--------|-------|
+ *  bdl[n] bdl[n+j]    bdl[e-1] | bdl[0] bdl[1] bdl[m]   bdl[n-2] bdl[n-1]
  */
 int hda_dsp_stream_setup_bdl(struct snd_sof_dev *sdev,
 			     struct snd_dma_buffer *dmab,
-			     struct hdac_stream *stream)
+			     struct hdac_stream *stream, u32 init_offset)
 {
 	struct sof_intel_dsp_bdl *bdl;
-	int i, offset, period_bytes, periods;
-	int remain, ioc;
+	unsigned int i, period_bytes, entries;
+	u32 bytes, offset, size = 0;
+	bool ioc, set_ioc, is_wrap = false;
 
 	period_bytes = stream->period_bytes;
 	dev_dbg(sdev->dev, "period_bytes:0x%x\n", period_bytes);
 	if (!period_bytes)
 		period_bytes = stream->bufsize;
 
-	periods = stream->bufsize / period_bytes;
+	entries = stream->bufsize / period_bytes;
 
-	dev_dbg(sdev->dev, "periods:%d\n", periods);
+	/* count the last offset fragment entry */
+	if (stream->bufsize % period_bytes)
+		entries++;
 
-	remain = stream->bufsize % period_bytes;
-	if (remain)
-		periods++;
+	/* count the last index fragment entry */
+	if (init_offset % period_bytes)
```

Comment:
@keyonjie if init_offset is == bufsize, would the number of entries be 2? shouldnt it be just 1?

Reply:
@ranj063 that won't happen, please refer to:
```
	sdev->trace_init_offset = (sdev->trace_init_offset +
				   sdev->trace_offset) %
				   sdev->dmatb.bytes;
```
init_offset is in range of [0, bufsize).

Reply:
got it!